{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import itertools\n",
    "import time\n",
    "import pickle\n",
    "from IPython.display import clear_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker = 'cytb' # co1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load file with species names, sequence IDs and coordinates, column names should be 'seq', 'species', 'y', 'x'\n",
    "coords = pd.read_csv('place file name here', sep='\\t')\n",
    "coords.dropna(inplace=True)\n",
    "mammsSP = set(coords.species)\n",
    "\n",
    "len(mammsSP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### function to calculate GD per species and spatial unit\n",
    "\n",
    "$ \widehat{\Pi} $=  \\frac{1}{n \\choose 2} \\sum_\\limits{i=1}^{n-1} \\sum_\\limits{j=i+1}^{n} \\frac{k_{ij}} {m_{ij}} $\n",
    "\n",
    "$n$ is the number of sequences of the species in the cell\n",
    "\n",
    "$k_{ij}$ is the number of different nucleotides between sequence i and sequence j\n",
    "\n",
    "$m_{ij}$ is the number of shared base pairs between sequence i and j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculateSpGD(fastaSegs, cellSeqs, includedSeqs, GDthrs = 1):\n",
    "    pairGD = []\n",
    "    pairDst = []\n",
    "    nOfSeqs = set()\n",
    "    newCellSeqs = [s for s in cellSeqs if s in ''.join(includedSeqs)]\n",
    "    sequenceComb = list(itertools.combinations(newCellSeqs, 2))\n",
    "    if len(sequenceComb) > 0:\n",
    "        for combo in sequenceComb:\n",
    "            seq1 = re.search('{}.*\\n(.*?)\\n'.format(combo[0]), fastaSegs).group(1)\n",
    "            seq2 = re.search('{}.*\\n(.*?)\\n'.format(combo[1]), fastaSegs).group(1)\n",
    "            siteList = list(zip(seq1, seq2))\n",
    "            overlapLen = len([i for i in siteList if '-' not in i and 'N' not in i])\n",
    "            if overlapLen >= 0.5*max([len(seq1.replace('-',  '').replace('N', '')), len(seq2.replace('-',  '').replace('N', ''))]): #if sequences overlap in at least x% of the longest sequence\n",
    "                NumberOfSNPs = sum([1 for i in siteList if '-' not in i and 'N' not in i and len(set(i)) > 1])\n",
    "                OnePairGD = NumberOfSNPs / overlapLen\n",
    "                if OnePairGD < GDthrs: #maximum GD per pair\n",
    "                    pairGD.append(OnePairGD)\n",
    "                    nOfSeqs.add(combo[0])\n",
    "                    nOfSeqs.add(combo[1])\n",
    "                    #get distance\n",
    "                    xc1 = coords.loc[combo[0]]['x']\n",
    "                    yc1 = coords.loc[combo[0]]['y']\n",
    "                    xc2 = coords.loc[combo[1]]['x']\n",
    "                    yc2 = coords.loc[combo[1]]['y']\n",
    "                    dst = getDistanceHarv(xc1, yc1, xc2, yc2, 6371)\n",
    "                    pairDst.append(dst)\n",
    "        if len(pairGD) == 0:\n",
    "            return np.nan, np.nan, []\n",
    "        else:\n",
    "            return pairGD, pairDst, nOfSeqs\n",
    "    else:\n",
    "        return np.nan, np.nan, []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Scan sequences that fall outside of species native ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#functions to calculate distances\n",
    "def deg2rad(deg):\n",
    "    return deg * (math.pi/180)\n",
    "\n",
    "def getDistanceHarv(lon1, lat1, lon2, lat2, earthRadius):\n",
    "    lo1, la1, lo2, la2 = radians(lon1), radians(lat1), radians(lon2), radians(lat2)\n",
    "    \n",
    "    R = earthRadius # Radius of the earth in km\n",
    "    dlon = deg2rad(lon2-lon1)\n",
    "    dlat = deg2rad(lat2-lat1)\n",
    "   \n",
    "    a = sin(dlat / 2)**2 + cos(la1) * cos(la2) * sin(dlon / 2)**2\n",
    "    c = 2 * atan2(sqrt(a), sqrt(1 - a))\n",
    "\n",
    "    d = R * c\n",
    "    return d;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from shapely.geometry import shape, MultiPolygon, mapping, Point, MultiPoint\n",
    "from shapely.ops import transform, nearest_points\n",
    "import fiona #import fiona after shapely otherwise buffer crashes\n",
    "import json\n",
    "from functools import partial\n",
    "import pyproj\n",
    "\n",
    "#origin\n",
    "#1 Native\n",
    "#2 Reintroduced\n",
    "#3 Introduced\n",
    "#4 Vagrant\n",
    "#5 Origin Uncertain \n",
    "\n",
    "#presence\n",
    "#1 Extant\n",
    "#2 Probably Extant\n",
    "#3 Possibly Extant\n",
    "#4 Possibly Extinct\n",
    "#5 Extinct (post 1500)\n",
    "#6 Presence Uncertain \n",
    "\n",
    "#load IUCN shape file with ranges\n",
    "mammRanges = fiona.open('TERRESTRIAL_MAMMALS.shp')\n",
    "##create dictionary with polygons for each species\n",
    "dSpPoly = {}\n",
    "for feature in mammRanges:\n",
    "    if feature['properties']['origin'] in [1,2] and feature['properties']['presence'] in [1,2]:\n",
    "        sp = list(feature['properties'].items())[1][1]\n",
    "        if feature['geometry']['type'] == 'MultiPolygon':\n",
    "            pls = list(shape(feature['geometry']))\n",
    "            dSpPoly.setdefault(sp.replace(' ', '_'), []).extend(pls)\n",
    "        else:\n",
    "            dSpPoly.setdefault(sp.replace(' ', '_'), []).extend([shape(feature['geometry'])])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "earthRadius = 6371\n",
    "\n",
    "inSeqs = []\n",
    "outSeqs = []\n",
    "\n",
    "distThrs = 385.9451 #km\n",
    "iseq = 0\n",
    "\n",
    "start_time = time.time()\n",
    "for seq,value in coords.iterrows():\n",
    "    iseq += 1\n",
    "    clear_output(wait=True)\n",
    "    print('evaluating seq {}/{}'.format(iseq, len(coords)))\n",
    "    x = value['x']\n",
    "    y = value['y']\n",
    "    sp = value['species']\n",
    "    if sp in dSpPoly:\n",
    "        #check if seq in mamm range\n",
    "        p = Point(float(x), float(y))\n",
    "        inter = sum([p.within(i) for i in dSpPoly[sp]])\n",
    "        if inter >= 1:\n",
    "            inSeqs.append(seq)\n",
    "        else:\n",
    "            #get nearest polygon edges to the point\n",
    "            nearPoints = [nearest_points(MultiPoint(dSpPoly[sp][i].exterior.coords), p)[0].coords[0] for i, pol in enumerate(dSpPoly[sp]) ]\n",
    "            #get minimum distance\n",
    "            mD = min([getDistanceHarv(tup[0], tup[1], float(x), float(y), earthRadius) for  tup in nearPoints])\n",
    "            if mD < distThrs:\n",
    "                inSeqs.append(seq)\n",
    "            else:\n",
    "                outSeqs.append(seq) #excluded sequences\n",
    "                \n",
    "    else:\n",
    "        syn = synon[synon.synonyms.str.contains(sp)]['species'].values \n",
    "        if len(syn) > 1 and syn[0] in dSpPoly:\n",
    "            #check if seq in mamm range\n",
    "            p = Point(float(x), float(y))\n",
    "            inter = sum([p.within(i) for i in dSpPoly[sp]])\n",
    "            if inter >= 1:\n",
    "                inSeqs.append(seq)\n",
    "            else:\n",
    "                #get nearest polygon edges to the point\n",
    "                nearPoints = [nearest_points(MultiPoint(dSpPoly[sp][i].exterior.coords), p)[0].coords[0] for i, pol in enumerate(dSpPoly[sp]) ]\n",
    "                #get minimum distance\n",
    "                mD = min([getDistanceHarv(tup[0], tup[1], float(x), float(y), earthRadius) for  tup in nearPoints])\n",
    "                if mD < distThrs:\n",
    "                    inSeqs.append(seq)\n",
    "                else:\n",
    "                    outSeqs.append(seq) #excluded sequences\n",
    "                    \n",
    "print('--- {} minutes ---\\n\\n'.format((time.time() - start_time)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate GD per species and cell and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### load python dictionary with cell id sequences contained\n",
    "sd = pickle.load(open( '{}/grid.p'.format(marker), \"rb\" ))    \n",
    "#new dictionary to store GD for mammals\n",
    "GDperSPmamms = {}\n",
    "\n",
    "start_time = time.time()\n",
    "totCells = len(sd)\n",
    "cellGD = {}\n",
    "\n",
    "for ci, cell in enumerate(sd):\n",
    "    clear_output(wait=True)\n",
    "    print('cell {} of {}'.format(ci+1, totCells))\n",
    "\n",
    "    mammalGD = []\n",
    "    NmammalSpecies = 0\n",
    "    NmammalSeqs = 0\n",
    "    amphGD = []\n",
    "    NamphSpecies = 0\n",
    "    NamphSeqs = 0\n",
    "\n",
    "    for species in sd[cell]:\n",
    "        # select species with at least 2 sequences\n",
    "        if len(sd[cell][species]) > 1:\n",
    "            if species in mammsSP:\n",
    "                #load alignment\n",
    "                with open('{0}/{1}.fasta'.format(marker, species)) as inf:\n",
    "                    data=inf.read()\n",
    "                    allPairGD, allPairDist, totSeqs = calculateSpGD(data, sd[cell][species], inSeqs)\n",
    "                    #store n of seqs and GD\n",
    "                    GDperSPmamms.setdefault(cell, {})[species] = {'nOfSeqs': totSeqs, 'GD': allPairGD, 'Dist': allPairDist}\n",
    "\n",
    "\n",
    "print('--- %s seconds ---\\n\\n\\n' % (time.time() - start_time))\n",
    "\n",
    "#save dictionary\n",
    "with open('{}/GD_cells.p'.format(marker), 'wb')) as fp:\n",
    "    pickle.dump(GDperSPmamms, fp, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate GD per species and latitudinal band"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### load python dictionary with band id and sequences contained\n",
    "sd = pickle.load(open( '{}/band.p'.format(marker), 'rb' ))\n",
    "#new dictionary to store GD for mammals\n",
    "GDperSPmammsBands = {}\n",
    "\n",
    "start_time = time.time()\n",
    "totCells = len(sd)\n",
    "bandGD = {}\n",
    "\n",
    "for bi, band in enumerate(sd):\n",
    "    clear_output(wait=True)\n",
    "    print('band {} of {}'.format(bi+1, totCells))\n",
    "\n",
    "    for species in sd[band]:\n",
    "        # select species with at least 2 sequences\n",
    "        if len(sd[band][species]) > 1:\n",
    "            if species in mammsSP:\n",
    "                #load alignment\n",
    "                with open('{0}/{1}.fasta'.format(marker, species)) as inf:\n",
    "                    data=inf.read()\n",
    "                    allPairGD, allPairDist, totSeqs = calculateSpGD(data, sd[band][species], inSeqs)\n",
    "                    #store n of seqs and GD\n",
    "                    GDperSPmammsBands.setdefault(band, {})[species] = {'nOfSeqs': totSeqs, 'GD': allPairGD, 'Dist': allPairDist}\n",
    "\n",
    "print('--- %s seconds ---\\n\\n\\n' % (time.time() - start_time))\n",
    "\n",
    "#save dictionary\n",
    "with open('{}/GD_bands.p'.format(marker), 'wb') as fp:\n",
    "    pickle.dump(GDperSPmammsBands, fp, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate GD per species and zoogeographic region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sd = pickle.load(open( '{}/wallace.p'.format(marker), 'rb' ))\n",
    "\n",
    "start_time = time.time()\n",
    "totRegions = len(sd)\n",
    "#new dictionary to store GD for mammals\n",
    "wallGD = {}\n",
    "\n",
    "for wi, regio in enumerate(sd):\n",
    "    print('Region {} of {}: {}'.format(wi+1, totRegions, regio))\n",
    "\n",
    "    for species in sd[regio]:\n",
    "        # select species with at least 2 sequences\n",
    "        if len(sd[regio][species]) > 1:\n",
    "            if species in mammsSP:\n",
    "                #load alignment\n",
    "                with open('{0}/{1}.fasta'.format(marker, species)) as inf:\n",
    "                    data=inf.read()\n",
    "                    allPairGD, allPairDist, totSeqs = calculateSpGD(data, sd[regio][species], inSeqs)\n",
    "                    #store n of seqs and GD\n",
    "                    wallGD.setdefault(regio, {})[species] = {'nOfSeqs': totSeqs, 'GD': allPairGD, 'Dist': allPairDist}\n",
    "                    \n",
    "#save dictionary\n",
    "with open('{}/GD_wallace.p'.format(marker), 'wb') as fp:\n",
    "    pickle.dump(GDperSPmammsBands, fp, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
